MGDAPI-4539 add gcp provider skeleton and creds req

cecobask opened the pull request:
## Overview

Jira: https://issues.redhat.com/browse/MGDAPI-4539

## Verification

- Checkout this [branch of Delorean](https://github.com/integr8ly/delorean/pull/283) and create a GCP CCS cluster:
```
make -f make/ocm.mk ocm/cluster.json BYOC=true CLOUD_PROVIDER=gcp
```
- After the cluster installation completes, set up the cluster, using CRO:
```
make cluster/prepare
```
- Run CRO locally:
```
make run
```
- In another tab, create new blobstorage/postgres/redis instance:
```
make cluster/seed/blobstorage PROVIDER=gcp
```
- With the current implementation, the only thing that we should expect is to have no errors reconciling `CredetialsRequest`. No cloud resources are expected to be created at this point because there's no implementation for that functionality yet.

valerymo commented on controllers/blobstorage/blobstorage_controller.go:
```diff
@@ -72,7 +73,11 @@ func New(mgr manager.Manager) (*BlobStorageReconciler, error) {
+	providerList := []providers.BlobStorageProvider{
+		openshift.NewBlobStorageProvider(client, logger),
+		awsBlobStorageProvider,
+		gcp.NewGCPBlobStorageProvider(client),
```
@cecobask , Maybe before using Openshift and GCP  BlobStorage Providers need check them, similar to AWS? (L72) --   awsBlobStorageProvider, err := aws.NewAWSBlobStorageProvider(client, logger))
 + error checking to be more informative - print provider name, in case of error

cecobask replied:
@valerymo,

`NewAWSBlobStorageProvider()`, `NewAWSPostgresProvider()` and `NewAWSRedisProvider()` are the only provider methods that can return errors when being initialized because the underlying [credential manager](https://github.com/cecobask/cloud-resource-operator/blob/MGDAPI-4539/pkg/providers/aws/credentials.go#L145-L163) is likely to throw an error if there was a problem with the STS secret.
The `openshift` and `gcp` providers don't throw errors when they're being initialized, therefore there is nothing to check for. Maybe in the future this will change and when it does, error checking will definitely be introduced.

valerymo commented on controllers/postgres/postgres_controller.go:
```diff
@@ -82,7 +83,11 @@ func New(mgr manager.Manager) (*PostgresReconciler, error) {
+	providerList := []providers.PostgresProvider{
+		openshift.NewOpenShiftPostgresProvider(client, clientSet, logger),
+		awsPostgresProvider,
+		gcp.NewGCPPostgresProvider(client),
```
Could you please look at previous comment  - to validate  Openshift and GCP BlobStorage Providers same way as awsPostgresProvider. And also - error to be informative.

valerymo commented on controllers/redis/redis_controller.go:
```diff
@@ -70,7 +71,11 @@ func New(mgr manager.Manager) (*RedisReconciler, error) {
+	providerList := []providers.RedisProvider{
+		openshift.NewOpenShiftRedisProvider(client, logger),
+		awsRedisProvider,
+		gcp.NewGCPRedisProvider(client),
```
same as two previouse comments - could we validate Openshift and GCP BlobStorage Providers same way as awsPostgresProvider. And also - error to be informative.

valerymo commented on doc/blobstorage.md:
````diff
@@ -4,7 +4,7 @@
 ```
 $ make cluster/prepare 
-$ make cluster/seed/blobstorage
+$ make cluster/seed/blobstorage PROVIDER=openshift
````
Could we add values for PROVIDER in case AWS and GCP?

valerymo commented on doc/postgresql.md:
````diff
@@ -4,7 +4,7 @@
 ```
 $ make cluster/prepare 
-$ make cluster/seed/postgres
+$ make cluster/seed/postgres PROVIDER=openshift
````
same as previous comment - Could we add values for PROVIDER in case AWS and GCP?

valerymo commented on doc/redis.md:
````diff
@@ -4,7 +4,7 @@
 ```
 $ make cluster/prepare 
-$ make cluster/seed/redis
+$ make cluster/seed/redis PROVIDER=openshift
````
same as previous comment - Could we add values for PROVIDER in case AWS and GCP?

valerymo commented on hack/postgres/README.md:
````diff
@@ -7,9 +7,9 @@ Scripts and helpers for working with postgres
 ```
 export ns=cloud-resources-load-test
-# Spin up workshop `postgres` instance
+# Spin up openshift `postgres` instance
````
similar to previous comments -  Could we somehow talk about two other Providers?

valerymo commented on hack/postgres/README.md:
```diff
@@ -60,7 +60,7 @@ oc exec $pod_name sh /var/lib/pgsql/memUsageExec.sh $aws_rds_db_host $aws_rds_db
 
 #### Clean Script
-Copy `clean.sh` file to provisioned postgres workshop pod  
+Copy `clean.sh` file to provisioned postgres openshift pod  
```
same as previous comment

valerymo commented on hack/redis/README.md:
````diff
@@ -17,10 +17,10 @@ Follow this guide to install [the latest version of oc](https://docs.openshift.c
 ```
 
-* Create a managed redis instance with `tier` set to `production` that you will run the load test against.
+* Create an AWS redis instance with `tier` set to `production` that you will run the load test against.
````
similar to previous comments

valerymo commented on pkg/client/util_test.go:
```diff
@@ -438,7 +438,7 @@ func TestReconcileRedis(t *testing.T) {
 				ctx:            context.TODO(),
 				client:         fake.NewFakeClientWithScheme(scheme),
-				deploymentType: "managed",
+				deploymentType: "aws",
```
@cecobask , maybe it's possible to do similar tests for GCP?

valerymo commented on pkg/providers/gcp/config.go:
```diff
@@ -0,0 +1,59 @@
+package gcp
```
@cecobask @laurafitzgerald  Could I ask , maybe we could have similar code/structure (as providers/gcp) for two other cases - AWS and Openshift (providers/aws, providers/openshift). Maybe not for this PR, but have follow-on task

cecobask replied:
The structure is mostly the same with all 3 providers. Not sure I follow what changes you are suggesting to do.
We can sync when you're back from PTO 👍🏻

valerymo replied:
@cecobask , yes, sorry, it looks good

valerymo reviewed (commented):
@cecobask Overall looks good, before running it. I placed few comments, could you please look.
Also, could you please provide more details how to create GCP CCS cluster (please see error below).  Thank you
[delorean (MGDAPI-4600)]$ make -f make/ocm.mk ocm/cluster.json BYOC=true CLOUD_PROVIDER=gcp
Error: GCP service account file not found
... Seems need file with GCP credentials ?

valerymo commented:
GCP seems a new area, so maybe it will be nice to add some references to GCP related documentation, and GCP setups steps need to be done (if any) before run delorean make command

laurafitzgerald commented on config/samples/cloud_resource_config.yaml:
```diff
@@ -3,7 +3,9 @@ apiVersion: v1
+    {"blobstorage":"openshift", "redis":"openshift", "postgres":"openshift"}
+  aws: |
     {"blobstorage":"aws","redis":"aws", "postgres":"aws"}
-  workshop: |
```
workshop is for installing in cluster. I don't think we need to remove it?

cecobask replied:
`workshop` isn't exactly removed. I made modification to the naming.
There used to be two deployment types: `workshop` and `managed`.
Since we're introducing GCP support, these names don't make sense anymore.
The deployment types are now:
- `workshop` -> `openshift`
- `managed` -> `aws`
- `gcp` - new type
This way the deployment types will match the provider names and reduce confusion.

valerymo commented:
I tried to run it both for AWS (regression, looks ok), and for GCP.
It's a skeleton, not all implemented yet for GCP, like Postgres reconciling, but what is done looks running ok.
Unit testing also good, but could be improved later, as not not all methods covered.
I will be in PTO next days, but if comments for code I placed before addressed, it will look good to me.
Thanks @cecobask , @laurafitzgerald

cecobask commented:
> I tried to run it both for AWS (regression, looks ok), and for GCP. It's a skeleton, not all implemented yet for GCP, like Postgres reconciling, but what is done looks running ok. Unit testing also good, but could be improved later, as not not all methods covered. I will be in PTO next days, but if comments for code I placed before addressed, it will look good to me. Thanks @cecobask , @laurafitzgerald

@valerymo,

I addressed all suggested feedback and added missing unit tests to cover the missing coverage.
When you're back from PTO, let me know if anything else is needed.

adam-cattermole commented on pkg/providers/config.go:
```diff
@@ -79,8 +79,9 @@ func (m *ConfigMapConfigManager) buildDefaultConfigMap() *v1.ConfigMap {
-			"workshop": "{\"blobstorage\":\"openshift\", \"redis\":\"openshift\", \"postgres\":\"openshift\"}",
+			"openshift": "{\"blobstorage\":\"openshift\", \"redis\":\"openshift\", \"postgres\":\"openshift\"}",
+			"aws":       "{\"blobstorage\":\"aws\", \"redis\":\"aws\", \"postgres\":\"aws\"}",
+			"gcp":       "{\"blobstorage\":\"gcp\", \"redis\":\"gcp\", \"postgres\":\"gcp\"}",
```
I agree that moving to this model makes more sense but my only concern would be if the config map already exists. If the provider config map is present with `managed` configuration what will happen when updating CRO?

adam-cattermole replied:
Ah looks like value in the CR takes precedence so will continue to use `managed` in CR until/unless we override it.

valerymo commented:
@cecobask , thank you for addressing comments. It looks good for me. I see that integration/feature branch is used now. Thanks.

valerymo commented:
/lgtm

laurafitzgerald commented on config/samples/integreatly_v1alpha1_redis.yaml:
```diff
@@ -16,3 +16,5 @@ spec:
   # set to true to force any pending service updates on the redis instance
   applyImmediately: false
+  # whether service updates should take place
+  maintenanceWindow: false
```
I think we could solve this with a rebase of the feature branch.

laurafitzgerald reviewed (commented):
Hey @cecobask seems we've got some additional changes into this pr from master. I'm okay with things merging to the feature branch and resolve it there. I think there's a bit of stubs and things here which would be helpful to have merged.

cecobask commented on Makefile:
```diff
@@ -4,12 +4,12 @@ IMAGE_NAME ?= cloud-resource-operator
 NAMESPACE=cloud-resource-operator
-PREV_VERSION=0.39.0
-VERSION=0.40.0
+PREV_VERSION=0.40.0
```
@laurafitzgerald, that's correct. Instead of doing a separate PR to rebase the feature branch with master, I pulled the commits from master to this branch, together with the work on the current JIRA. Is that okay?

laurafitzgerald replied:
yeah that's fine. it's a bit easier to review when rebasing but i think it's fine this time. are you ready to go ahead and merge this pr then?

cecobask replied:
It's ready to be merged 👍🏻 Just needs to be approved. Would you mind posting your approval?
